vm/src/stdlib/builtins.rs

```diff
@@ -248,11 +248,35 @@ mod builtins {
     #[cfg(feature = "rustpython-compiler")]
     #[pyfunction]
     fn eval(
-        source: Either<PyStrRef, PyRef<crate::builtins::PyCode>>,
+        source: Either<
+            Either<PyStrRef, crate::builtins::PyBytesRef>,
+            PyRef<crate::builtins::PyCode>,
+        >,
         scope: ScopeArgs,
         vm: &VirtualMachine,
     ) -> PyResult {
-        run_code(vm, source, scope, compile::Mode::Eval, "eval")
+        // source as string
+        let code = match source {
+            Either::A(either) => {
+                let source: &[u8] = match either {
+                    Either::A(ref a) => a.as_str().as_bytes(),
+                    Either::B(ref b) => b.as_bytes(),
+                };
+                for x in source {
+                    if *x == 0 {
+                        return Err(vm.new_value_error(
+                            "source code string cannot contain null bytes".to_owned(),
+                        ));
+                    }
+                }
+
+                Ok(Either::A(
+                    vm.ctx.new_str(std::str::from_utf8(source).unwrap()),
```

Comment:
what happens if `bytes` include non-ascii bytes?
what happens if `bytes` include invalid utf8 character?

Reply:
> what happens if bytes include invalid utf8 character?

It causes `panic`. I tested with the below python code:

```python
eval(b'\xff')
```

And it prints:


```text
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 0, error_len: Some(1) }', vm/src/stdlib/builtins.rs:268:64
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Because the `C` language `const char *` type for string and also byte array, it can be represented as a single type with a compiler option. So I'll see more the compiler, parser section. If you have useful documentations to recommend, please leave them as comments 🙇🏻‍♂️ 

```c
// _PyPegen_run_parser_from_string
    if (flags != NULL && flags->cf_flags & PyCF_IGNORE_COOKIE) {
        tok = PyTokenizer_FromUTF8(str, exec_input);
    } else {
        tok = PyTokenizer_FromString(str, exec_input);
    }
```

Reply:
What I thought was `.unwrap()` need to be avoided because CPython would not panic but raise  a `SyntaxError`.
The major difference between FromUTF8 and FromString is using decoder or not. The decode error seems to be wrapped by SyntaxError somewhere.

Reply:
Ah, I agree with your thought. 🙏🏻  I'll work for wrapping the decode error.

Reply:
When you execute `eval(b'\xff')` in CPython implementation, you will see the below output:

```
>>> eval(b"\xff")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1
    �
    ^
SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

I set `SyntaxError: (unicode error) 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte` message as format and I tried to apply it in [714ce4d](https://github.com/RustPython/RustPython/pull/4052/commits/714ce4d07d7c900e7ceb2dab9eae725f15f749e4).